addons/minio/template/base/install.sh

```diff
@@ -10,10 +10,290 @@ function minio_pre_init() {
+    cp -Rfp "$bkp_location/.kurl" "$MINIO_HOSTPATH/"
+}
+
+# minio_uses_fs_format verifies if minio uses the legacy fs format. greps the file /data/.minio.sys/format.json
+# from within the pod.
+function minio_uses_fs_format() {
+    local format_string
+    format_string=$(kubectl exec -n $MINIO_NAMESPACE deploy/minio -- cat /data/.minio.sys/format.json 2>/dev/null)
+    if [ -z "$format_string" ]; then
+        bail "failed to read /data/.miniosys/format.json inside minio pod"
+    fi
+    echo "$format_string" | grep -q '"format":"fs"'
+}
+
+# minio_svc_has_no_endpoints validates that the provided service in the provided namespace has no endpoints.
+function svc_has_no_endpoints() {
+    namespace=$1
+    service=$2
+    kubectl get endpoints --no-headers -n "$namespace" "$service" -o custom-columns=:subsets |grep -q none
+}
+
+# disable_minio_svc disables the minio service by tweaking its service selector.
+function disable_minio_svc() {
+    kubectl patch svc -n $MINIO_NAMESPACE minio -p '{"spec":{"selector":{"app":"does-not-exist"}}}'
+}
+
+# enable_minio_svc sets the minio service selector back to its default.
+function enable_minio_svc() {
+    kubectl patch svc -n $MINIO_NAMESPACE minio -p '{"spec":{"selector":{"app":"minio"}}}'
+}
+
+# prepare_volumes_for_migration sets the retention policy of the volumes to "retain" for both minio and its migration deployment.
+function prepare_volumes_for_migration() {
+    # get the pv in use by the minio deployment.
+    local minio_pv
+    minio_pv=$(kubectl get pvc --no-headers -n $MINIO_NAMESPACE minio-pv-claim -o custom-columns=:.spec.volumeName)
+    if [ -z "$minio_pv" ]; then
+        return 1
+    fi
+
+    # get the pv in use by the fs migration deployment.
+    local migration_pv
+    migration_pv=$(kubectl get pvc --no-headers -n $MINIO_NAMESPACE minio-migrate-fs-backend-pv-claim -o custom-columns=:.spec.volumeName)
+    if [ -z "$migration_pv" ]; then
+        return 1
+    fi
+
+    # set both of the pv retention policies to 'retain'.
+    kubectl patch pv "$minio_pv" -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+    kubectl patch pv "$migration_pv" -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
+}
+
+# migrate_fs_backend migrates a minio running with 'fs' backend into the new default backend type (xl). spawns a new
+# minio deployment then copies the object buckets from the old deployment into the new one. this function generates
+# unavailability of the original minio service.
+function migrate_fs_backend() {
+    # if minio is not running then we can't migrate its data. minio may be deployed by means of a statefulset as well
+    # but on this case the running pods are guaranteed to not leverage the fs backend.
+    local minio_replicas
+    minio_replicas=$(kubectl get --no-headers deploy minio -n $MINIO_NAMESPACE -o custom-columns=:.spec.replicas 2>/dev/null)
+    if [ -z "$minio_replicas" ] || [ "$minio_replicas" = "0" ]; then
```

Comment:
is `[ -z "$minio_replicas" ]` an error? should we bail? what about `[ "$minio_replicas" = "0" ]`? is that expected? should there be two separate cases?

Reply:
The Minio deployment [may not](https://github.com/replicatedhq/ekco/blob/main/pkg/cluster/minio.go#L160) exist so a failure there is somehow expected in later Minio releases. What if we report an error if it exists but the number of replicas is zero ? I am very curious to hear your thoughts on the matter.